runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -222,6 +230,47 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
     return Optional.<Boolean>absent();
   }
 
+  /**
+   * Check if all data watermark value for each step of the pipeline reach to max.
+   *
+   * <p>If throwing IOException while retrieving job metrics, ignore and return false.
+   */
+  boolean atMaxWatermark(DataflowPipelineJob job) {
+    JobMetrics metrics;
+    try {
+      metrics = options.getDataflowClient().projects().jobs()
```

Comment:
With checkForSuccess, your potentially getting the metrics twice for each iteration. Please refactor to do the metrics call only once.